app/services/cap_update_service.rb

```diff
@@ -0,0 +1,34 @@
+class CapUpdateService
+  attr_accessor :school
+
+  def initialize(args = {})
+    @school = args[:school]
+  end
+
+  def update!(order_state:, cap:)
+    update_order_state!(order_state)
+    allocation = update_cap!(cap)
+
+    notify_computacenter!(allocation.id)
+  end
+
+private
+
+  def update_order_state!(order_state)
+    @school.update!(order_state: order_state)
+  end
+
+  def update_cap!(cap)
+    allocation = SchoolDeviceAllocation.find_or_initialize_by(school_id: @school.id, device_type: 'std_device')
+    # we only take the cap from the user if they chose specific circumstances
+    # for both other states, we need to infer a new cap from the chosen state
+    allocation.cap = allocation.cap_implied_by_order_state(order_state: @school.order_state, given_cap: cap)
+    allocation.save!
```

Comment:
This line blows up if the allocation is invalid (this issue already existed before this PR but it feels like now is a good time to fix it).

This can happen since #469 introduced some validations that are more sophisticated than what `Support::EnableOrdersForm` catches (because the form object is very simplistic).

I can see a few approaches to fix this:

1. Make the validations on `SchoolDeviceAllocation` and `Support::EnableOrdersForm` consistent. It feels like this approach would introduce unnecessary duplication between those models.
2. Move the construction/fetch of the allocation to `Support::EnableOrdersForm`, and validate that the allocation is valid inside `EnableOrdersForm`. This approach enables copying the validation errors from the allocation onto `EnableOrdersForm`, making it easy to display the error messages on the page.
3. Call `allocation.valid?` before trying to `save!` and skip the call to the Computacenter API if the return is false. That way the controller can also check for validity of the allocation and display the errors if it's not.

Reply:
On standup/dev huddle just now, we agreed to deal with this in a separate PR

Reply:
1. feels like the best approach to me. Maybe we can refactor the validation to a separate validator class, and then call it from both places

2. I did think about _not_ introducing a service object, but instead moving the save functionality to the form object like they have in Apply. I decide against, on the grounds of generally preferring form objects to represent a form and having any complex back-end interactions handled at the service layer  (or model layer directly if they're not so complex)

3. `save!` will raise an error if the record is invalid, then the error will bubble out to the controller and the transaction will be rolled back. The question then is how best for the form to indicate the validation error to the user - but if we do 1., then that problem is solved.

The only way I can see that the allocation could be invalid is if the user chooses the specific circumstances, and then enters a cap that's greater than the allocation - that's a problem with the form data, and should be validated at the form level.

Reply:
I'm fairly relaxed about how we solve this, if we go for 1 and use a validator that's probably going to work well. May be worth combining the several validation rules into one custom validator, since they're all very similar anyway.